src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+---
+section: System
+chapter: Foundations
+title: Director
+description: Introduction to the Director and how to use it to build behaviours.
+slug: /system/foundations/director
+---
+
+The NUbots use the director system to control all their behaviour elements.
+It allows composable behaviours that can take advantage of transition rules to ensure that the robot performs the right task at the right time.
+
+The Director does this by providing a set of new DSL keywords to NUClear that describe what tasks to run, what conditions are required, and how those tasks build on other tasks.
+It then uses these keywords to make a graph that it will use to decide which tasks to execute in order to achive the goals of the modules.
+
+The two DSL components that are added are Providers and Tasks.
+
+## Providers
+
+Providers are implemented as NUClear reactions and live in a NUClear `on` statement.
+They are the components which "provide" the functionality needed to perform a task, either by doing it directly or by running a series of subtasks.
+They form the basis of the modules that implement behaviours.
+Each provider will provide for a specific message type in a specific situation.
+For example a provider which executed a walk task, the code might look like:
+
+```cpp
+on<Provides<Walk>>().then(...);
+```
+
+Provider are triggered whenever a task needs to be executed by the director, however they can also be triggered using other NUClear keywords.
+The director will ensure that any provider which is not currently active will not run if it is not active, even if the other triggers would make it run.
+For example if the walk engine module wants to update at 10Hz when it is running a provider may look like.
+
+```
+on<Provides<Walk>, Every<10, Per<std::chrono::seconds>>().then(...);
+```
+
+This provider would then run once every time the walk command was updated, as well as running at 10Hz while the provider is active.
```

Comment:
The way it disables providers is not so much like calling .disable() on the reaction, it will be using the same mechanism where if you have an on trigger with a with, and the with hasn't been emitted yet it won't run.

So really when you have something else (like an every) running on a provider that the director is not using, it's more like there isn't a task there for it to access so it doesn't run.

Reply:
Oh right. So you could disable these reactors as well as deactivating the providers?